### PR TITLE
Adapts selenium tests to che multiuser changes

### PR DESCRIPTION
**Pull request by akorneta**
### What does this PR do?
Adapts selenium tests to che multiuser changes.
Adds keycloak auth client to allow get and refresh access tokens for test users.
Adds tests startup option:
`--multiuser`
Note that without this parameter, selenium tests will proceed the same flow as on single user che.

Adds configuration properties:
Defines default test user for che single / multiuser selenium tests
`che.test_user.email`
`che.test_user.password`

Defines admin test user for che multiuser selenium tests
`che.admin_user.email`
`che.admin_user.password`

### What issues does this PR fix or reference?
https://github.com/eclipse/che/issues/5930

### CODE REVIEW

**Comment by tolusha** [comment]
Is ready to review?

**Comment by akorneta** [comment]
@tolusha yes, in general, things related to che multiuser almost done except the additional user creation.
Another part that related to launch che single user will be done in this pr asap.

File: selenium/che-selenium-core/src/main/java/org/eclipse/che/selenium/core/requestfactory/TestAdminHttpJsonRequestFactory.java
Comment (dmytro-ndp) [inline selenium/che-selenium-core/src/main/java/org/eclipse/che/selenium/core/requestfactory/TestAdminHttpJsonRequestFactory.java]: It's Che-specific settings which should be re-written in tests for Codenvy onprem/saas. 
I propose to create **TestAdminEmailProvider** and **TestAdminPasswordProvider** interfaces and next implementations for them:
- **CheTestAdminEmailProvider** and **CheTestAdminPasswordProvider** implementations in selenium tests for Che (where _che.admin_user.email_ and _che.admin_user.password_ config properties are using);
- **CodenvyTestAdminEmailProvider** and **CodenvyTestAdminPasswordProvider** implementations for Codenvy (where _codenvy.admin.name_, _codenvy.admin.initial_password_, and _codenvy.admin.password_ config properties are used similarly to the class https://github.com/codenvy/codenvy/blob/master/selenium/codenvy-selenium-test/src/main/java/com/codenvy/selenium/core/user/OnpremAdminTestUser.java)

File: selenium/che-selenium-core/src/main/java/org/eclipse/che/selenium/core/requestfactory/TestAdminHttpJsonRequestFactory.java
Comment (dmytro-ndp) [inline selenium/che-selenium-core/src/main/java/org/eclipse/che/selenium/core/requestfactory/TestAdminHttpJsonRequestFactory.java]: Let's log admin's email here as well so as it could be different.

File: selenium/che-selenium-core/src/main/java/org/eclipse/che/selenium/core/requestfactory/TestDefaultUserHttpJsonRequestFactory.java
Comment (dmytro-ndp) [inline selenium/che-selenium-core/src/main/java/org/eclipse/che/selenium/core/requestfactory/TestDefaultUserHttpJsonRequestFactory.java]: It's better to log user's email for the full information so as user's name is usual just the part of email before "@"

File: selenium/che-selenium-core/src/main/java/org/eclipse/che/selenium/core/requestfactory/TestDefaultUserHttpJsonRequestFactory.java
Comment (dmytro-ndp) [inline selenium/che-selenium-core/src/main/java/org/eclipse/che/selenium/core/requestfactory/TestDefaultUserHttpJsonRequestFactory.java]: we can use **testUser.getAuthToken()** here, can't we?

File: selenium/che-selenium-core/src/main/java/org/eclipse/che/selenium/core/requestfactory/TestDefaultUserHttpJsonRequestFactory.java
Comment (dmytro-ndp) [inline selenium/che-selenium-core/src/main/java/org/eclipse/che/selenium/core/requestfactory/TestDefaultUserHttpJsonRequestFactory.java]: This class looks like redundant after default test user is replaced by regular test user, doesn't it?

File: selenium/che-selenium-core/src/main/java/org/eclipse/che/selenium/core/requestfactory/TestUserHttpJsonRequestFactory.java
Comment (dmytro-ndp) [inline selenium/che-selenium-core/src/main/java/org/eclipse/che/selenium/core/requestfactory/TestUserHttpJsonRequestFactory.java]: we can use **testUser.getAuthToken()** here, can't we?

File: selenium/che-selenium-core/src/main/java/org/eclipse/che/selenium/core/requestfactory/TestUserHttpJsonRequestFactory.java
Comment (dmytro-ndp) [inline selenium/che-selenium-core/src/main/java/org/eclipse/che/selenium/core/requestfactory/TestUserHttpJsonRequestFactory.java]: It's better to log user's email for the full information so as user's name is usual just the part of email before "@"

File: selenium/che-selenium-core/src/main/java/org/eclipse/che/selenium/core/requestfactory/TestUserHttpJsonRequestFactory.java
Comment (akorneta) [reply selenium/che-selenium-core/src/main/java/org/eclipse/che/selenium/core/requestfactory/TestUserHttpJsonRequestFactory.java]: Well, we need to guarantee that returned access token is valid and not expired so it's better to use authServiceClient.login call to get/refresh/login user's token

File: selenium/che-selenium-core/src/main/java/org/eclipse/che/selenium/core/requestfactory/TestDefaultUserHttpJsonRequestFactory.java
Comment (akorneta) [reply selenium/che-selenium-core/src/main/java/org/eclipse/che/selenium/core/requestfactory/TestDefaultUserHttpJsonRequestFactory.java]: yes you are right

File: selenium/che-selenium-core/src/main/java/org/eclipse/che/selenium/core/user/TestUserFactory.java
Comment (dmytro-ndp) [inline selenium/che-selenium-core/src/main/java/org/eclipse/che/selenium/core/user/TestUserFactory.java]: What is the propose of this method? 
It looks as redundant so as simple `@Inject TestUser test user;` could be used instead of testUserFactory.create()

File: selenium/che-selenium-core/src/main/java/org/eclipse/che/selenium/core/user/TestUserImpl.java
Comment (dmytro-ndp) [inline selenium/che-selenium-core/src/main/java/org/eclipse/che/selenium/core/user/TestUserImpl.java]: Did you test it? 
It will lead to an error of injection when we try to create random user by the command 
`@Inject TestUser testUser;`.

File: selenium/che-selenium-core/src/main/java/org/eclipse/che/selenium/core/user/TestUserFactory.java
Comment (akorneta) [reply selenium/che-selenium-core/src/main/java/org/eclipse/che/selenium/core/user/TestUserFactory.java]: This method needs in case when in tests there need to be some user with generated email and password. So for codenvy I planed to bind the `TestUser` to implementation wich call this method.

File: selenium/che-selenium-core/src/main/java/org/eclipse/che/selenium/core/requestfactory/TestAdminHttpJsonRequestFactory.java
Comment (akorneta) [reply selenium/che-selenium-core/src/main/java/org/eclipse/che/selenium/core/requestfactory/TestAdminHttpJsonRequestFactory.java]: added

File: selenium/che-selenium-test/src/main/java/org/eclipse/che/selenium/core/TestUserServiceClientProvider.java
Comment (dmytro-ndp) [inline selenium/che-selenium-test/src/main/java/org/eclipse/che/selenium/core/TestUserServiceClientProvider.java]: Looks like a useless command. IMHO it's better to explicitly define proper user service clients in **CheSeleniumSingleUserModule** and **CheSeleniumMultiUserModule**.

File: selenium/che-selenium-core/src/main/java/org/eclipse/che/selenium/core/client/TestUserServiceClient.java
Comment (dmytro-ndp) [inline selenium/che-selenium-core/src/main/java/org/eclipse/che/selenium/core/client/TestUserServiceClient.java:39]: We need this method for https://github.com/codenvy/codenvy/blob/master/selenium/codenvy-selenium-test/src/test/java/com/codenvy/selenium/factory/AuthenticateAndAcceptFactoryThroughGitHubOAuthTest.java#L89-L89

Comment (akorneta) [reply selenium/che-selenium-core/src/main/java/org/eclipse/che/selenium/core/client/TestUserServiceClient.java:39]: I'll adapt this code in codenvy PR

File: selenium/che-selenium-test/src/test/resources/conf/selenium.properties
Comment (dmytro-ndp) [inline selenium/che-selenium-test/src/test/resources/conf/selenium.properties]: I can't find the usage of these properties. Can you, please, write an example of usage?

File: selenium/che-selenium-core/src/main/java/org/eclipse/che/selenium/core/user/TestUserImpl.java
Comment (akorneta) [reply selenium/che-selenium-core/src/main/java/org/eclipse/che/selenium/core/user/TestUserImpl.java]: yes for usecase described [here](https://github.com/eclipse/che/pull/6456#discussion_r141295205) it works pretty fine

File: selenium/che-selenium-test/src/test/resources/conf/selenium.properties
Comment (akorneta) [reply selenium/che-selenium-test/src/test/resources/conf/selenium.properties]: I'm working on it and will notify when it will be done

File: selenium/che-selenium-test/src/test/resources/suites/CheSuite.xml
Comment (dmytro-ndp) [inline selenium/che-selenium-test/src/test/resources/suites/CheSuite.xml]: It have not been finally decided to remove support of SVN plugin yet https://github.com/eclipse/che/issues/5484#issuecomment-328075154

File: selenium/che-selenium-test/src/test/resources/suites/PassingSuite.xml
Comment (dmytro-ndp) [inline selenium/che-selenium-test/src/test/resources/suites/PassingSuite.xml]: What is this suite for? We have already had CheSuite.xml, and it is using by default.

**Review by dmytro-ndp — CHANGES_REQUESTED** [review]
IMHO we need to ensure that these changes don't break execution of selenium tests for Single User Che, Codenvy onprem and Codenvy SaaS.

File: selenium/che-selenium-core/src/main/java/org/eclipse/che/selenium/core/user/TestUserFactory.java
Comment (dmytro-ndp) [reply selenium/che-selenium-core/src/main/java/org/eclipse/che/selenium/core/user/TestUserFactory.java]: I see. But we don't need assisted injection, that is TestUserFactory, for creation an instance from constructor without **@Assisted** parameters at all. 
@Inject annotation is able to provide required instance of TestUser.

File: selenium/che-selenium-core/src/main/java/org/eclipse/che/selenium/core/requestfactory/CheAdminHttpJsonRequestFactory.java
Comment (dmytro-ndp) [inline selenium/che-selenium-core/src/main/java/org/eclipse/che/selenium/core/requestfactory/CheAdminHttpJsonRequestFactory.java]: Usually we add **Test** word into the name of selenium test framework classes to avoid confusion with Che product codebase.

File: selenium/che-selenium-test/src/test/resources/conf/selenium.properties
Comment (akorneta) [reply selenium/che-selenium-test/src/test/resources/conf/selenium.properties]: @dmytro-ndp  I deleted these properties because for the first step of integration selenium tests with  che multiuser it's overhead.

File: selenium/che-selenium-test/src/test/resources/suites/PassingSuite.xml
Comment (akorneta) [reply selenium/che-selenium-test/src/test/resources/suites/PassingSuite.xml]: removed

**Comment by akorneta** [comment]
Fixed single user che test launching, for codenvy and saas assemblies there will be separate prs. Removed additional che user properties.
@tolusha @dmytro-ndp please take a look.

File: selenium/che-selenium-test/selenium-tests.sh
Comment (tolusha) [inline selenium/che-selenium-test/selenium-tests.sh:31]: Make it simpler. Just export what you need.

File: selenium/che-selenium-test/src/main/java/org/eclipse/che/selenium/core/factory/TestFactory.java
Comment (tolusha) [inline selenium/che-selenium-test/src/main/java/org/eclipse/che/selenium/core/factory/TestFactory.java:39]: You moved TestFactory into the test module. Unfortunately we can't mix page objects and selenium core part. 
Describe usecase.

File: selenium/che-selenium-test/src/test/resources/suites/CheSuite.xml
Comment (tolusha) [reply selenium/che-selenium-test/src/test/resources/suites/CheSuite.xml]: @akorneta Pls, restore removed tests.

File: selenium/che-selenium-test/src/test/resources/conf/selenium.properties
Comment (tolusha) [inline selenium/che-selenium-test/src/test/resources/conf/selenium.properties]: It is n't clear if user has to be predefined in single user assembly. Could you clarify pls

File: selenium/che-selenium-core/src/main/java/org/eclipse/che/selenium/core/requestfactory/TestUserHttpJsonRequestFactory.java
Comment (dmytro-ndp) [reply selenium/che-selenium-core/src/main/java/org/eclipse/che/selenium/core/requestfactory/TestUserHttpJsonRequestFactory.java]: But **testUser.getAuthToken()** can refresh user's token every time we call it, can't it?

File: selenium/che-selenium-test/src/test/resources/conf/selenium.properties
Comment (dmytro-ndp) [reply selenium/che-selenium-test/src/test/resources/conf/selenium.properties]: I see. We can add it in future on demand.

File: selenium/che-selenium-test/selenium-tests.sh
Comment (dmytro-ndp) [reply selenium/che-selenium-test/selenium-tests.sh:31]: @tolusha: we should remove **--multiuser** parameter from list of parameters because of **webdriver.sh** doesn't recognize it and throws an error.

File: selenium/che-selenium-core/src/main/java/org/eclipse/che/selenium/core/user/TestUserFactory.java
Comment (akorneta) [reply selenium/che-selenium-core/src/main/java/org/eclipse/che/selenium/core/user/TestUserFactory.java]: removed method + changed annotation

File: selenium/che-selenium-test/src/test/resources/conf/selenium.properties
Comment (akorneta) [reply selenium/che-selenium-test/src/test/resources/conf/selenium.properties]: added default values, so for single user che it's not needed to provide default user data

File: selenium/che-selenium-core/src/main/java/org/eclipse/che/selenium/core/requestfactory/TestUserHttpJsonRequestFactory.java
Comment (tolusha) [inline selenium/che-selenium-core/src/main/java/org/eclipse/che/selenium/core/requestfactory/TestUserHttpJsonRequestFactory.java:39]: Can we login once and then return authentication token?

File: selenium/che-selenium-test/selenium-tests.sh
Comment (tolusha) [reply selenium/che-selenium-test/selenium-tests.sh:31]: That's what I told. In case of multi use assembly:
```
export CHE_MULTIUSER=true
```

And then check environment variable in the code.

File: selenium/che-selenium-test/src/main/java/org/eclipse/che/selenium/core/CheSeleniumMultiUserModule.java
Comment (tolusha) [inline selenium/che-selenium-test/src/main/java/org/eclipse/che/selenium/core/CheSeleniumMultiUserModule.java]: There is mess with class names.
`CheTest`, `Test`, `TestChe`

File: selenium/che-selenium-test/src/main/java/org/eclipse/che/selenium/core/CheSeleniumSuiteModule.java
Comment (tolusha) [inline selenium/che-selenium-test/src/main/java/org/eclipse/che/selenium/core/CheSeleniumSuiteModule.java:99]: I can see the same bindings in both CheSeleniumMultiUserModule and CheSeleniumSingleUserModule.
Let's keep them in CheSeleniumSuiteModule

File: selenium/che-selenium-test/selenium-tests.sh
Comment (dmytro-ndp) [reply selenium/che-selenium-test/selenium-tests.sh:31]: We should remove _--multiuser_ parameter from parameters because otherwise it will be passed to **webdriver.sh** as it and will cause an error at start of execution of selenium tests:
> (target/bin/webdriver.sh "$TESTS_SCOPE" **$@**)

File: selenium/che-selenium-core/src/main/java/org/eclipse/che/selenium/core/requestfactory/TestUserHttpJsonRequestFactory.java
Comment (dmytro-ndp) [reply selenium/che-selenium-core/src/main/java/org/eclipse/che/selenium/core/requestfactory/TestUserHttpJsonRequestFactory.java:39]: We need to update token regularly because of little default token timeout around 5 minutes.

File: selenium/che-selenium-test/src/main/java/org/eclipse/che/selenium/core/CheSeleniumMultiUserModule.java
Comment (dmytro-ndp) [reply selenium/che-selenium-test/src/main/java/org/eclipse/che/selenium/core/CheSeleniumMultiUserModule.java]: Let me explain to Anton the naming scheme:
1. Usually we add **Test** prefix to the name of classes which are specific to selenium tests like the following:
> AuthServiceClient -> **Test**AuthServiceClient
> HttpJsonRequestFactory -> **Test**UserHttpJsonRequestFactory
2. In case of different implementation for Che, Codenvy Onprem, Keycloak, Codenvy SaaS, we add name of product at start of class name:
> TestAuthServiceClient -> **Che**TestAuthServiceClient
or -> **Onpem**TestAuthServiceClient
or -> **Keycloak**TestAuthServiceClient

According to this scheme, correct name of class 
> Test**Che**DefaultUserHttpJsonRequestFactory > **Che**TestDefaultUserHttpJsonRequestFactory.

Comment (akorneta) [reply selenium/che-selenium-test/src/main/java/org/eclipse/che/selenium/core/CheSeleniumMultiUserModule.java]: ok I'll fix that

File: selenium/che-selenium-test/src/main/java/org/eclipse/che/selenium/core/factory/TestFactoryInitializer.java
Comment (dmytro-ndp) [inline selenium/che-selenium-test/src/main/java/org/eclipse/che/selenium/core/factory/TestFactoryInitializer.java]: typo

**Comment by dmytro-ndp** [comment]
@akorneta: thank you for quick fixup.
I would appropriate it if commit messages will reflect content of changes.

**Comment by tolusha** [comment]
There are 2 requirements to merge into master branch:
* don't mix page objects with core part of selenium tests
* ensure running tests for codenvy/saas projects

**Comment by tolusha** [comment]
It is ok to be merged into multiuser branch and unblock platform team workking

**Review by tolusha — APPROVED** [review]

**Comment by skabashnyuk** [comment]
@tolusha  we do our best to met your requirements before merge che-multiuser branch in master

**Review by skabashnyuk — APPROVED** [review]

**Pull request merged**
